baseplate/secrets/fetcher.py

```diff
@@ -108,16 +111,73 @@ def ttl_to_time(ttl):
     """Factory that makes authenticated clients."""
-    def __init__(self, base_url, role):
+    def __init__(self, base_url, role, auth_type, auth_path):
         self.base_url = base_url
         self.role = role
+        self.auth_type = auth_type
+        self.auth_path = auth_path
         self.session = requests.Session()
         self.client = None
 
     def _make_client(self):
-        """Authenticate with Vault and return a client containing that token.
+        """Obtain a client token from an auth backend and return a Vault client with it."""
+        if self.auth_type == "aws_ec2":
+            client_token, lease_duration = self._vault_aws_ec2_auth()
+        elif self.auth_type == "kubernetes":
+            client_token, lease_duration = self._vault_kubernetes_auth()
+        else:
+            raise ValueError("Unsupported vault auth_type '%s' provided" % self.auth_type)
 
-        This authenticates with Vault as a specified role using its AWS EC2
+        return VaultClient(
+            self.session,
+            self.base_url,
+            client_token,
+            lease_duration,
+        )
+
+    def _vault_kubernetes_auth(self):
+        """This authenticates with Vault as a specified role using its
+        Kubernetes auth backend. This involves sending Vault a JSON Web Token
+        associated with a Kubernetes service account mounted at a well-known
+        location within a running pod. Vault should be configured with a
+        mapping binding roles to corresponding Kubernetes service accounts and
+        namespaces along with appropriate policies. For example, a pod running
+        in the `prod` namespace with the service account name `my-server`
+        requires a Vault configuration created like so:
+
+            vault write /auth/kubernetes/cluster-name/role/my-server-role \
+                bound_service_account_names=my-server \
+                bound_service_account_namespaces=prod \
+                policies=my-servers-policies \
+                max_ttl=4h
+
+        See https://www.vaultproject.io/docs/auth/kubernetes.html for more info.
+
+        """
+        try:
+            with open(K8S_SERVICE_ACCOUNT_TOKEN_FILE, "r") as f:
+                token = f.read()
+        except IOError:
+            logger.error("Could not read Kubernetes token file '%s'",
+                         K8S_SERVICE_ACCOUNT_TOKEN_FILE)
+            raise
+
+        login_data = {
+            "jwt": token,
+            "role": self.role,
+        }
+
+        logger.debug("Obtaining Vault token via kubernetes auth.")
+        response = self.session.post(
+            urljoin(self.base_url, self.auth_path),
+            json=login_data,
+        )
+        response.raise_for_status()
+        auth = response.json()["auth"]
+        return auth["client_token"], ttl_to_time(auth["lease_duration"])
+
+    def _vault_aws_ec2_auth(self):
+        """This authenticates with Vault as a specified role using its AWS EC2
```

Comment:
[pep-257] asks for: "Multi-line docstrings consist of a summary line just like a one-line docstring, followed by a blank line, followed by a more elaborate description.". this line should keep being a summary like it was before rather than running into the rest of the first paragraph.

[pep-257]: https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings